stellargraph/core/element_data.py

```diff
@@ -331,28 +358,41 @@ def __init__(self, shared, type_starts):
 
     def _init_directed_adj_lists(self):
         # record the edge ilocs of incoming, outgoing and both-direction edges
-        in_dict = {}
-        out_dict = {}
+        in_dict = dict((node, []) for node in np.sort(pd.unique(self.targets)))
```

Comment:
I think this is close to what is potentially a more optimal way to do this:

```python
def rle(x):
    """
    run-length encode x, e.g. [1, 1, 1, 1, 3, 4, 4] -> [1, 3, 4], [4, 1, 2]
    """
    ...

sorted_source_edge_ilocs = np.argsort(sources)
sorted_sources = sources[sorted_source_edge_ilocs]

run_node_iloc, run_length = rle(sorted_sources)
# handle isolated nodes, which won't appear in `run_node_iloc`
lengths = np.zeros(number_of_nodes)
lengths[run_node_iloc] = run_length

splits = np.cumsum(lengths)
flat = sorted_source_edge_ilocs
```

I may've got the details wrong, but this is attempting to reproduce the `flat` and `splits` values in `FlatAdjacencyList `.

`rle` might be something like https://stackoverflow.com/a/32681075/1256624.

Reply:
nice! I went with this approach and sped up the construction time of the adj lists by ~6x!